frontend/src/diagram/DropArea.tsx

```diff
@@ -0,0 +1,172 @@
+/*******************************************************************************
+ * Copyright (c) 2019, 2020 Obeo.
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Obeo - initial API and implementation
+ *******************************************************************************/
+import { useMutation } from '@apollo/client';
+import { canDrop } from 'diagram/toolServices';
+import { useProject } from 'project/ProjectProvider';
+import PropTypes from 'prop-types';
+import React, { useCallback } from 'react';
+import { invokeDropToolOnDiagramMutation } from './operations';
+
+const DATA_DIAGRAM_ID = 'data-diagramid';
+
+const DATA_DESCRIPTION_ID = 'data-descriptionid';
+
+const DATA_NODE_ID = 'data-nodeid';
+
+const SVG_NAMESPACE = 'http://www.w3.org/2000/svg';
+
+const toolType = PropTypes.shape({
+  id: PropTypes.string.isRequired,
+  label: PropTypes.string.isRequired,
+  imageURL: PropTypes.string,
+});
+const propTypes = {
+  representationId: PropTypes.string.isRequired,
+  toolSections: PropTypes.arrayOf(
+    PropTypes.shape({
+      label: PropTypes.string.isRequired,
+      tools: PropTypes.arrayOf(toolType).isRequired,
+      defaultTool: toolType.isRequired,
+    })
+  ).isRequired,
+  invokeHover: PropTypes.func.isRequired,
+  children: PropTypes.node.isRequired,
+};
+
+export const DropArea = ({ representationId, toolSections, invokeHover, children }) => {
+  const { id } = useProject();
+
+  const [invokeDropTool] = useMutation(invokeDropToolOnDiagramMutation);
+
+  const dropElement = useCallback(
+    (tool: any, objectId: string, diagramElementId?: string) => {
+      const { id: toolId } = tool;
+      const input = {
+        projectId: id,
+        representationId,
+        objectId,
+        toolId,
+      };
+      if (diagramElementId) {
+        input['diagramElementId'] = diagramElementId;
+      }
+
+      const variables = {
+        input,
+      };
+      invokeDropTool({ variables });
+    },
+    [id, representationId, invokeDropTool]
+  );
+
+  const searchId = useCallback((dom) => {
+    if (dom.hasAttribute(DATA_DIAGRAM_ID)) {
+      return dom.getAttribute(DATA_DIAGRAM_ID);
+    } else if (dom.hasAttribute(DATA_NODE_ID)) {
+      return dom.getAttribute(DATA_NODE_ID);
+    } else if (dom.parentElement) {
+      const parentDom = dom.parentElement;
+      if (parentDom.namespaceURI === SVG_NAMESPACE) {
+        return searchId(parentDom);
+      }
+    }
+    return null;
+  }, []);
+
+  const searchDescriptionId = useCallback((dom) => {
```

Comment:
Why all those usage of useCallback?

Reply:
To avoid to re-create a function at each DropArea re-render.
It makes sense for me but it is troubling to you, I will delete them.